core/trino-main/src/main/java/io/trino/sql/planner/optimizations/ArrayAggregationAfterDistinctOptimizer.java

```diff
@@ -0,0 +1,106 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.optimizations;
+
+import com.google.common.collect.ImmutableMap;
+import io.trino.Session;
+import io.trino.execution.warnings.WarningCollector;
+import io.trino.metadata.Metadata;
+import io.trino.operator.aggregation.arrayagg.ArrayAggregationFunction;
+import io.trino.operator.scalar.ArrayDistinctFunction;
+import io.trino.sql.planner.PlanNodeIdAllocator;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.SymbolAllocator;
+import io.trino.sql.planner.TypeProvider;
+import io.trino.sql.planner.plan.AggregationNode;
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.ProjectNode;
+import io.trino.sql.planner.plan.SimplePlanRewriter;
+import io.trino.sql.tree.FunctionCall;
+
+import static java.util.Objects.requireNonNull;
+
+public class ArrayAggregationAfterDistinctOptimizer
+        implements PlanOptimizer
+{
+    private final Metadata metadata;
+
+    public ArrayAggregationAfterDistinctOptimizer(Metadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        return SimplePlanRewriter.rewriteWith(new ArrayAggregationAfterDistinctOptimizer.Optimizer(metadata, idAllocator), plan);
+    }
+
+    private static class Optimizer
+            extends SimplePlanRewriter<Void>
+    {
+        private final Metadata metadata;
+        private final PlanNodeIdAllocator idAllocator;
+
+        private Optimizer(Metadata metadata, PlanNodeIdAllocator idAllocator)
+        {
+            this.metadata = metadata;
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitProject(ProjectNode node, RewriteContext<Void> context)
+        {
+            PlanNode rewritten = context.defaultRewrite(node);
+            if (rewritten instanceof ProjectNode &&
+                    ((ProjectNode) rewritten).getSource() instanceof AggregationNode &&
+                    ((ProjectNode) rewritten).getAssignments().getExpressions().stream()
+                            .anyMatch(f -> f instanceof FunctionCall &&
+                                    metadata.decodeFunction(((FunctionCall) f).getName()).getSignature().getName().equals(ArrayDistinctFunction.NAME)) &&
```

Comment:
You need to check that `array_distinct` in the ProjectNode is applied to the right `array_agg` in the AggregationNode. For now, we only know that there is some `array_distinct` and some `array_agg`. They should be identified and processed pairwise. Consider this plan:

```
ProjectNode:
  a <- array_distinct(x)
  b <- array_distinct(z)
  c <- array_distinct(y)
        |
     AggregationNode (group by z):
       y <- array_agg(p)
       x <- array_agg(q)
```
After the rewrite there should be:
```
ProjectNode:
  a <- x
  b <- array_distinct(z)
  c <- y
        |
     AggregationNode (group by z):
       y <- array_agg(DISTINCT p)
       x <- array_agg(DISTINCT q)
```